compiler/rustc_middle/src/ty/structural_impls.rs

```diff
@@ -728,7 +711,18 @@ impl<'tcx> TypeSuperFoldable<TyCtxt<'tcx>> for ty::Const<'tcx> {
         folder: &mut F,
     ) -> Result<Self, F::Error> {
         let ty = self.ty().try_fold_with(folder)?;
-        let kind = self.kind().try_fold_with(folder)?;
+        let kind = match self.kind() {
+            ConstKind::Param(p) => ConstKind::Param(p.try_fold_with(folder)?),
+            ConstKind::Infer(i) => ConstKind::Infer(i.try_fold_with(folder)?),
+            ConstKind::Bound(d, b) => {
+                ConstKind::Bound(d.try_fold_with(folder)?, b.try_fold_with(folder)?)
+            }
+            ConstKind::Placeholder(p) => ConstKind::Placeholder(p.try_fold_with(folder)?),
+            ConstKind::Unevaluated(uv) => ConstKind::Unevaluated(uv.try_fold_with(folder)?),
+            ConstKind::Value(v) => ConstKind::Value(v.try_fold_with(folder)?),
+            ConstKind::Error(e) => ConstKind::Error(e.try_fold_with(folder)?),
+            ConstKind::Expr(e) => ConstKind::Expr(e.try_fold_with(folder)?),
+        };
```

Comment:
why isn't this an impl on the generic `ConstKind` with the appropriate bounds?

Reply:
I just copied how  the typefoldable stuff works for `Ty`. which is that we have the impls on `Ty` and `TyKind` isnt `TypeVisitable` or `TypeFoldable`. It's a change from the status quo but it doesn't really seem like we acutally used `ConstKind: TypeVisitable/TypeFoldable` anywhere except the `Const: TypeVisitable/TypeFoldable` impl and I rather have `Const` work the same as `Ty` where possible.

Reply:
I suspected as much, makes sense, we can revisit if more usages ever come up